go/tools/gazelle/merger/merger.go

```diff
@@ -0,0 +1,232 @@
+/* Copyright 2016 The Bazel Authors. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package merger provides methods for merging parsed BUILD files.
+package merger
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"sort"
+
+	bzl "github.com/bazelbuild/buildifier/core"
+)
+
+var (
+	mergeableFields = map[string]bool{
+		"srcs": true,
+		"deps": true,
+	}
+)
+
+// MergeWithExisting looks for an existing BUILD file at file.Path
+// loads it, and attempts to merge elements of newfile into it.
+// returns newfile, nil if FileNotExists
+func MergeWithExisting(newfile *bzl.File) (*bzl.File, error) {
+	b, err := ioutil.ReadFile(newfile.Path)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return newfile, nil
+		}
+		return nil, err
+	}
+	f, err := bzl.Parse(newfile.Path, b)
+	if err != nil {
+		return nil, err
+	}
+
+	var newStmt []bzl.Expr
+	for _, s := range newfile.Stmt {
+		c, ok := s.(*bzl.CallExpr)
+		if !ok {
+			return nil, fmt.Errorf("got %v expected only CallExpr in %q", s, newfile.Path)
+		}
+		other, err := match(f, c)
+		if err != nil {
+			return nil, err
+		}
+		if other == nil {
+			newStmt = append(newStmt, c)
+			continue
+		}
+		if literal(c.X) == "load" {
+			mergeLoad(c, other, f)
+		} else {
+			merge(c, other)
+		}
+	}
+	f.Stmt = append(f.Stmt, newStmt...)
+	return f, nil
+}
+
+// merge takes new info from src and merges into dest.
+// pre: these calls are the same X and 'name'
+func merge(src, dest *bzl.CallExpr) {
+	for _, item := range src.List {
+		nkv, ok := item.(*bzl.BinaryExpr)
+		if !ok {
+			continue
+		}
+		k := literal(nkv.X)
+		if !mergeableFields[k] {
+			continue
+		}
+		// TODO(pmbethe09): allow '# keep' on src files.
```

Comment:
I have thought about this, but wanted to get the baseline in before working on it.
For example, in buildifier we have the generated file "parse.y.go" which will need a # keep before gazelle will work correctly.